client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-recipe.ts

```diff
@@ -1,34 +1,88 @@
+import { useSearchParams } from 'react-router-dom';
 import { ONBOARD_STORE } from '../../../../../stores';
 import { decodePatternId } from '../utils';
 import type { Pattern, Category } from '../types';
 import type { OnboardSelect } from '@automattic/data-stores';
-import type { Design } from '@automattic/design-picker/src/types';
 import type { GlobalStylesObject } from '@automattic/global-styles';
 
 const useRecipe = ( siteId = 0, patterns: Pattern[], categories: Category[] ) => {
-	const incrementIndexRef = useRef( 0 );
-	const [ header, setHeader ] = useState< Pattern | null >( null );
-	const [ footer, setFooter ] = useState< Pattern | null >( null );
-	const [ sections, setSections ] = useState< Pattern[] >( [] );
-	const [ colorVariation, setColorVariation ] = useState< GlobalStylesObject | null >( null );
-	const [ fontVariation, setFontVariation ] = useState< GlobalStylesObject | null >( null );
+	const [ searchParams, setSearchParams ] = useSearchParams();
+
+	const header_pattern_id = searchParams.get( 'header_pattern_id' );
+	const footer_pattern_id = searchParams.get( 'footer_pattern_id' );
+
+	const section_pattern_ids = ( searchParams.get( 'pattern_ids' ) || '' )
+		.split( ',' )
+		.filter( Boolean );
+
+	const color_variation_title = searchParams.get( 'color_variation_title' );
+	const font_variation_title = searchParams.get( 'font_variation_title' );
+
+	const patternsById = keyBy( patterns, 'ID' );
+	const categoriesByName = keyBy( categories, 'name' );
+
+	const header = header_pattern_id ? patternsById[ decodePatternId( header_pattern_id ) ] : null;
+	const footer = footer_pattern_id ? patternsById[ decodePatternId( footer_pattern_id ) ] : null;
+
+	/* Adds a unique "key" prop to each section pattern,
+	   so that when the patterns are rendered in a list,
+	   they are not re-rendered unnecessarily when the list
+	   is manipulated (some elements get appended / removed).
+	*/
+	const [ keyedSections, setKeyedSections ] = useState< Pattern[] >( [] );
+
+	/* To sync the keyed sections with the search params,
+	   we first initialize the sections from the search params.
+	*/
+	const initialSections = useMemo(
+		() =>
+			section_pattern_ids
+				.map( ( patternId ) => patternsById[ decodePatternId( patternId ) ] )
+				.filter( Boolean )
+				.map( ( pattern: Pattern ) => {
+					const [ firstCategory ] = Object.keys( pattern.categories );
+					const category = categoriesByName[ firstCategory ];
+					return {
+						...pattern,
+						category,
+					};
+				} ),
+		/* We initialize the sections (at most) once when the patterns and categories are finally loaded
+		   (hence the ".length"s in the dependency array).
+
+		   We are ignoring the changes from the search params (section_pattern_ids),
+	       since after this point, the changes will be handled by setSections().
+	    */
+		[ patterns.length, categories.length ] // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
@arthur791004 unfortunately we cannot depend on just `patternsById`, even if we use `useMemo()` on it, because `patterns` itself changes every render 🥲

Reply:
I guess we can do it below and it also avoids converting patterns/categories to the Map on every rendering. WDYT?

```js
const patternsById = useMemo( () => keyBy( patterns, 'ID' ), [ patterns.length ] );
const categoriesByName = useMemo( () => keyBy( categories, 'name' ), [ categories.length ] );
```

Reply:
Oh that's cleaner, updated, thanks 👍